src/main/java/org/folio/services/transactions/BatchTransactionService.java

```diff
@@ -82,33 +83,46 @@ private Future<Void> updateTransaction(Transaction transaction, RequestContext r
-      .map(connected -> {
-        if (FALSE.equals(connected)) {
+    // Usually it is not allowed to delete an encumbrance connected to an approved invoice.
+    // To avoid adding a dependency to mod-invoice, we check if there is a related awaitingPayment transaction.
+    // It is OK to delete an encumbrance connected to a *cancelled* invoice *if* the batch includes a change to the
+    // matching pending payment to remove the link to the encumbrance.
+    String query = convertIdsToCqlQuery(ids, "awaitingPayment.encumbranceId", "==", " OR ");
+    return getTransactions(query, requestContext)
+      .map(pendingPayments -> {
+        if (pendingPayments.isEmpty()) {
           return null;
         }
-        logger.warn("validateDeletion:: Tried to delete transactions but one is connected to an invoice, ids={}",
-          batch.getIdsOfTransactionsToDelete());
-        throw new HttpException(422, DELETE_CONNECTED_TO_INVOICE.toError());
+        ids.forEach(id -> {
+          Optional<Transaction> existingPP = pendingPayments.stream()
+            .filter(pp -> id.equals(pp.getAwaitingPayment().getEncumbranceId())).findFirst();
+          if (existingPP.isEmpty()) {
+            return;
+          }
+          if (TRUE.equals(existingPP.get().getInvoiceCancelled())) {
+            Optional<Transaction> matchingPPInBatch = batch.getTransactionsToUpdate().stream()
+              .filter(pp -> existingPP.get().getId().equals(pp.getId())).findFirst();
+            if (matchingPPInBatch.isPresent() && matchingPPInBatch.get().getAwaitingPayment().getEncumbranceId() == null) {
+              return;
+            }
+          }
+          logger.warn("validateDeletion:: Tried to delete transactions but one is connected to an invoice, id={}", id);
+          throw new HttpException(422, DELETE_CONNECTED_TO_INVOICE.toError());
+        });
+        return null;
       });
   }
 
-  private Future<Boolean> anyConnectedToInvoice(List<String> ids, RequestContext requestContext) {
-    // We want to know if the order with the given encumbrance is connected to an invoice.
-    // To avoid adding a dependency to mod-invoice, we check if there is a related awaitingPayment transaction
-    String query = convertIdsToCqlQuery(ids, "awaitingPayment.encumbranceId", "==", " OR ");
-    return getTransactions(query, requestContext)
-      .map(collection -> collection.getTotalRecords() > 0);
-  }
-
-  private Future<TransactionCollection> getTransactions(String query, RequestContext requestContext) {
+  private Future<List<Transaction>> getTransactions(String query, RequestContext requestContext) {
     var requestEntry = new RequestEntry(resourcesPath(TRANSACTIONS))
       .withOffset(0)
       .withLimit(Integer.MAX_VALUE)
       .withQuery(query);
-    return restClient.get(requestEntry.buildEndpoint(), TransactionCollection.class, requestContext);
+    return restClient.get(requestEntry.buildEndpoint(), TransactionCollection.class, requestContext)
```

Comment:
thinking we need some change here. 
If for example someone tryis to delete for example 200 transactions, in this case quey to check awaitingPayment.encumbranceId will be so long  and can be more than 4kb that is max for GET request.
So need to use batching as in similar places in this module.

Reply:
Agreed.